src/performance.js

```diff
@@ -58,8 +66,48 @@ export class Performance {
 
     /** @const @private {!Array<TickEventDef>} */
     this.events_ = [];
+
+    /** @private {boolean} */
+    this.isViewerVisible_ = false;
+
+    this.win.addEventListener('unload', this.flush.bind(this));
   }
 
+  /**
+   * Gets primary amp core services.
+   * @return {!Performance}
+   */
+  init() {
+    this.viewer_ = viewerFor(this.win);
+    this.resources_ = resourcesFor(this.win);
+    return this;
+  }
+
+  /**
+   * Listens to viewer and resource events.
+   */
+  listen() {
+    if (!(this.viewer_ && this.resources_)) {
+      return;
+    }
+
+    // User requested document to be visible
+    this.viewer_.whenFirstVisible().then(() => {
+      this.isViewerVisible_ = true;
```

Comment:
yep, agree. i have some local code that reads it in the `init` and @cramforce has corrected my logic so ill be making more changes here.